examples/tv-casting-app/tv-casting-common/core/CastingPlayerDiscovery.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "CastingPlayerDiscovery.h"
+
+namespace matter {
+namespace casting {
+namespace core {
+
+using namespace chip::System;
+using namespace chip::Controller;
+using namespace chip::Dnssd;
+using namespace std;
+
+void DeviceDiscoveryDelegateImpl::OnDiscoveredDevice(const chip::Dnssd::DiscoveredNodeData & nodeData)
+{
+
+    VerifyOrReturn(mClientDelegate != nullptr,
+                   ChipLogError(NotSpecified, "CastingPlayerDeviceDiscoveryDelegate, mClientDelegate is a nullptr"));
+
+    // convert nodeData to CastingPlayer
+    Attributes attributes;
+    strcpy(attributes.id, nodeData.resolutionData.hostName);
+
+    char port[kPortMaxLength] = {};
+    snprintf(port, sizeof(port), "%u", nodeData.resolutionData.port);
+    strcat(attributes.id, port);
+
+    strcpy(attributes.name, nodeData.commissionData.deviceName);
+    strcpy(attributes.host_name, nodeData.resolutionData.hostName);
+    attributes.num_IPs = nodeData.resolutionData.numIPs;
+    for (unsigned j = 0; j < attributes.num_IPs; j++)
+    {
+        attributes.ip_address[j] = nodeData.resolutionData.ipAddress[j];
+    }
+    attributes.product_id = nodeData.commissionData.productId;
+    attributes.vendor_id  = nodeData.commissionData.vendorId;
+    attributes.type       = nodeData.commissionData.deviceType;
+
+    Strong<CastingPlayer> player = std::make_shared<CastingPlayer>(attributes);
+
+    std::vector<Strong<CastingPlayer>> * castingPlayers = CastingPlayerDiscovery::GetInstance()->GetCastingPlayers();
```

Comment:
While this works, it is not ideal, because we are returning a pointer to the otherwise vector of CastingPlayers from GetCastingPlayers(). That vector is intended to be private to CastingPlayerDiscovery so some random code from outside that class cannot modify it. 

A nicer solution would be to declare DeviceDiscoveryDelegateImpl as a friend class to CastingPlayerDiscovery and let GetCastingPlayers return a copy of the Vector of CastingPlayers (not a pointer to it). Then DeviceDiscoveryDelegateImpl will have direct access to modify castingPlayers, but some other random code will not.

Reply:
updated.